### PR TITLE
Refactor services and handlers. And fixed bug of getting all categories.

### PR DESCRIPTION
**Pull request by immmus**
https://github.com/odpi/egeria/commit/0202181c24f90061a589b09f5d220200da8730a0 - Held refactoring in handlers. Removed duplicate parameters from constructors 
and removed unused and ported methods to handlers.

https://github.com/odpi/egeria/commit/a1ff64100ba36bc37fd7b385164bf1b0d28895c6 - Solved the problem of getting all categories using sequencingProperty and sequencingOrder. 

https://github.com/odpi/egeria/commit/7174b755186e777954f5831a5d03dc54c070d8c3 - Removed unused constants.

### CODE REVIEW

**Comment by davidradl** [comment]
@immmus please could include the @number of the issue that this pr refers to. The issue should only document what is being fixed. Please could you describe what the bug is and the associated fix you are proposing in the new issue. If there is a change to the API, please ensure you have updated the subject area FVT to test that the change works.  

We are branching this week so we are not putting anything non urgent into master, until we have branched.

**Comment by immmus** [comment]
> @immmus please could include the @Number of the issue that this pr refers to. The issue should only document what is being fixed. Please could you describe what the bug is and the associated fix you are proposing in the new issue. If there is a change to the API, please ensure you have updated the subject area FVT to test that the change works.
> 
> We are branching this week so we are not putting anything non urgent into master, until we have branched.


This pr refers to problem # 1 described here - https://github.com/odpi/egeria/issues/3090 . I supplemented description. Yes I conducted these tests and also I checked all actions manually to finally make sure that they work correctly. I also took into account your wish in our conversation and made using sequencingProperty and sequencingOrder, adding this to all method.

The problem No. 2 with the presence of unused methods in services is also solved. While researching the code, I realized that these methods were transferred to handlers, but not removed from the services. Therefore, I decided to remove them.

**Comment by immmus** [comment]
Other changes it my sentence for enhancement code.

**Comment by immmus** [comment]
I think that these changes must be here. - https://github.com/odpi/egeria/pull/3137/commits/9e56a27076ed15ad8827d477f270c7c5f9800992

@planetf1 Yes, I running these tests -> `open-metadata-test/open-metadata-fvt/access-services-fvt/subject-area-fvt/`

**Comment by immmus** [comment]
@davidradl When me to wait for approve this pr?;)

**Comment by davidradl** [comment]
> @davidradl When me to wait for approve this pr?;)

Hi @immmus, I would like to get https://github.com/odpi/egeria/issues/3112 in first - this is another large subject area refactor, that effects all the API layers. If you like you could review it. I am hoping to merge it on Wednesday unless there is any adverse feedback. I will review this once the code base has been updated.

**Comment by davidradl** [comment]
@immmus this pr is now conflicting with master - please could you resolve the conflicts.

**Comment by immmus** [comment]
> @immmus this pr is now conflicting with master - please could you resolve the conflicts.

@davidradl Hi, I resolved these conflicts.

File: open-metadata-implementation/access-services/subject-area/subject-area-server/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/utilities/SubjectAreaUtils.java [hunk 1]
```diff
@@ -103,130 +101,14 @@ public static void addStringToInstanceProperty(String key, String value, Instanc
         instanceProperties.setProperty(key, primitivePropertyValue);
     }
 
-    /**
-     * Set icon summaries from related media relationships by issuing a call to omrs using the related media guid - which is at one end of the relationship.
```
[the remark this answers is not in the document]
Comment (immmus) [reply open-metadata-implementation/access-services/subject-area/subject-area-server/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/utilities/SubjectAreaUtils.java:107]: All this code has been moved to another class so as not to have dependencies in the parameters. I didn’t delete anything, I just removed an unnecessary parameter
https://github.com/odpi/egeria/pull/3137/files#diff-45f8a331380b8ce75ae5ea724cc52d46

File: open-metadata-implementation/access-services/subject-area/subject-area-server/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/handlers/SubjectAreaCategoryHandler.java [hunk 2]
```diff
@@ -57,22 +52,10 @@
...
      */
-    public SubjectAreaCategoryHandler(String serviceName,
-                                      String serverName,
-                                      InvalidParameterHandler invalidParameterHandler,
-                                      OMRSRepositoryHelper repositoryHelper,
-                                      RepositoryHandler repositoryHandler,
-                                      OMRSAPIHelper oMRSAPIHelper,
-                                      RepositoryErrorHandler errorHandler) {
-        super(serviceName, serverName, invalidParameterHandler, repositoryHelper, repositoryHandler, oMRSAPIHelper);
+    public SubjectAreaCategoryHandler(OMRSAPIHelper oMRSAPIHelper) {
```
[the remark this answers is not in the document]
Comment (immmus) [reply open-metadata-implementation/access-services/subject-area/subject-area-server/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/handlers/SubjectAreaCategoryHandler.java:57]: There are no side effects in this case, I just took these general parameters to one place, removing duplication.

**Comment by immmus** [comment]
> Please could you add FVT tests for this . Please change CategoryFVT after it has added the first 2 categories then add this code ` // check that a null criteria finds all categories List<Category> results = findCategories(null); if (results.size() != 2) { throw new SubjectAreaFVTCheckedException("ERROR: Expected 2 back on the find got " + results.size()); }`
> 
> Please add the same tests for TermFVT and ProjectFVT. It would be good to add to SubjectAreaDefinitionCategoryFVT as well, but there are not as many tests there - but it would check that is can find classified GlossaryCategories and it comes back as expected.

**Comment by immmus** [comment]
> Please could you add FVT tests for this . Please change CategoryFVT after it has added the first 2 categories then add this code ` // check that a null criteria finds all categories List<Category> results = findCategories(null); if (results.size() != 2) { throw new SubjectAreaFVTCheckedException("ERROR: Expected 2 back on the find got " + results.size()); }`
> 
> Please add the same tests for TermFVT and ProjectFVT. It would be good to add to SubjectAreaDefinitionCategoryFVT as well, but there are not as many tests there - but it would check that is can find classified GlossaryCategories and it comes back as expected.

For good, all this functionality needs to be covered with Unit and Mock tests.

**Comment by davidradl** [comment]
> > Please could you add FVT tests for this . Please change CategoryFVT after it has added the first 2 categories then add this code ` // check that a null criteria finds all categories List<Category> results = findCategories(null); if (results.size() != 2) { throw new SubjectAreaFVTCheckedException("ERROR: Expected 2 back on the find got " + results.size()); }`
> > Please add the same tests for TermFVT and ProjectFVT. It would be good to add to SubjectAreaDefinitionCategoryFVT as well, but there are not as many tests there - but it would check that is can find classified GlossaryCategories and it comes back as expected.
> 
> For good, all this functionality needs to be covered with Unit and Mock tests.

@immmus We have a variety of testing approaches on the Egeria code base. For the areas I own I have down some junits and mocking for more complicated parts of the code. The risk of pervasive unit testing is that minor changes in the code can result in major junit rewrites. At the same time I have had good experiences of test driven development. For the Subject Area I know that running this code end to end covers and the main API functionality; for me the Subject Area API works as much as the FVT tests it. I would like FVT to be enhanced so this functionality does not inadvertently get  regressed. I hope that is ok with you.

**Comment by immmus** [comment]
> > > Please could you add FVT tests for this . Please change CategoryFVT after it has added the first 2 categories then add this code ` // check that a null criteria finds all categories List<Category> results = findCategories(null); if (results.size() != 2) { throw new SubjectAreaFVTCheckedException("ERROR: Expected 2 back on the find got " + results.size()); }`
> > > Please add the same tests for TermFVT and ProjectFVT. It would be good to add to SubjectAreaDefinitionCategoryFVT as well, but there are not as many tests there - but it would check that is can find classified GlossaryCategories and it comes back as expected.
> > 
> > 
> > For good, all this functionality needs to be covered with Unit and Mock tests.
> 
> @immmus We have a variety of testing approaches on the Egeria code base. For the areas I own I have down some junits and mocking for more complicated parts of the code. The risk of pervasive unit testing is that minor changes in the code can result in major junit rewrites. At the same time I have had good experiences of test driven development. For the Subject Area I know that running this code end to end covers and the main API functionality; for me the Subject Area API works as much as the FVT tests it. I would like FVT to be enhanced so this functionality does not inadvertently get regressed. I hope that is ok with you.

Well, I will add these tests, but only in my next pr, which will be released in 1-2 days. Please confirm this so that I do not have to face conflicts. I did a great job. Tests will be ready but slightly later). I just simply wait 16 days already and have gone far ahead in my current changes.

File: open-metadata-implementation/access-services/subject-area/subject-area-server/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/handlers/SubjectAreaCategoryHandler.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (davidradl) [reply open-metadata-implementation/access-services/subject-area/subject-area-server/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/handlers/SubjectAreaCategoryHandler.java:57]: @immmus Can you detail where the serviceName is supplied - I did not spot that?

**Comment by davidradl** [comment]
> Well, I will add these tests, but only in my next pr, which will be released in 1-2 days. Please confirm this so that I do not have to face conflicts. I did a great job. Tests will be ready but slightly later). I just simply wait 16 days already and have gone far ahead in my current changes.

@immmus OK I am happy for the tests to be done in another pr. I have the one one review comment I would like you to explain - if that is fine I will merge. I do appreciate your work. The 16 days was due to ongoing refactors and other commitments. I prioritised this as soon as I could; I apologise for the delay.

File: open-metadata-implementation/access-services/subject-area/subject-area-server/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/handlers/SubjectAreaCategoryHandler.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (immmus) [reply open-metadata-implementation/access-services/subject-area/subject-area-server/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/handlers/SubjectAreaCategoryHandler.java:57]: Yes, I brought all this logic to OMRSAPIHelper to get rid of duplication in the classes that use it
https://github.com/odpi/egeria/pull/3137/files#diff-45f8a331380b8ce75ae5ea724cc52d46

**Comment by immmus** [comment]
> > Well, I will add these tests, but only in my next pr, which will be released in 1-2 days. Please confirm this so that I do not have to face conflicts. I did a great job. Tests will be ready but slightly later). I just simply wait 16 days already and have gone far ahead in my current changes.
> 
> @immmus OK I am happy for the tests to be done in another pr. I have the one one review comment I would like you to explain - if that is fine I will merge. I do appreciate your work. The 16 days was due to ongoing refactors and other commitments. I prioritised this as soon as I could; I apologise for the delay.

Okey thx), I answered for your question)

File: open-metadata-implementation/access-services/subject-area/subject-area-server/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/handlers/SubjectAreaCategoryHandler.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (immmus) [reply open-metadata-implementation/access-services/subject-area/subject-area-server/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/handlers/SubjectAreaCategoryHandler.java:57]: get can be done using getters

[the remark this answers is not in the document]
Comment (davidradl) [reply open-metadata-implementation/access-services/subject-area/subject-area-server/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/handlers/SubjectAreaCategoryHandler.java:57]: @immmus  Sorry - I still do not get it. These are parameters supplied on a constructor - I am not sure how getters can set them? Please can you indicate the line numbers you are referring to.

**Comment by immmus** [comment]
> Please could you add FVT tests for this . Please change CategoryFVT after it has added the first 2 categories then add this code ` // check that a null criteria finds all categories List<Category> results = findCategories(null); if (results.size() != 2) { throw new SubjectAreaFVTCheckedException("ERROR: Expected 2 back on the find got " + results.size()); }`
> 
> Please add the same tests for TermFVT and ProjectFVT. It would be good to add to SubjectAreaDefinitionCategoryFVT as well, but there are not as many tests there - but it would check that is can find classified GlossaryCategories and it comes back as expected.

I already added these test on my next pr. For Glossary and Projects have already been

File: open-metadata-implementation/access-services/subject-area/subject-area-server/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/handlers/SubjectAreaCategoryHandler.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (immmus) [reply open-metadata-implementation/access-services/subject-area/subject-area-server/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/handlers/SubjectAreaCategoryHandler.java:57]: I meant that if you need parameters in the future, you can get them using the getter.
For example line 72 - https://github.com/odpi/egeria/blob/1e4eb3dcbed3029a179131801b4942842f47eb07/open-metadata-implementation/access-services/subject-area/subject-area-server/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/utilities/OMRSAPIHelper.java

Initialization as before,  I just removed the repetitions -https://github.com/odpi/egeria/pull/3137/files#diff-2d5677098874d14f512101635e153031

Initialization logic has not changed from this

[the remark this answers is not in the document]
Comment (davidradl) [reply open-metadata-implementation/access-services/subject-area/subject-area-server/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/handlers/SubjectAreaCategoryHandler.java:57]: @immmus I looked more closely and I think the logic is sound. I have merged.

[the remark this answers is not in the document]
Comment (immmus) [reply open-metadata-implementation/access-services/subject-area/subject-area-server/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/handlers/SubjectAreaCategoryHandler.java:57]: @davidradl Thank you:)

**Pull request merged**